R/dv8.R

```diff
@@ -0,0 +1,228 @@
+  cells_indices <- 1:nrow(depends_table)
+
+
+  # get the cell from each row in the depends table
+  getCell <- function(i){
+    depends_src <- depends_table[["src"]][[i]]
+    depends_dest <- depends_table[["dest"]][[i]]
+
+    # Look at first value in parse_depends table, get matching index in vars array
+    src_index <- variables_indices[depends_src] - 1
+    dest_index <- variables_indices[depends_dest] - 1
+
+    values = list()
+    parameters = list("Call", "Import", "Use", "Parameter", "Contain", "Create", "Extend",
+                      "Return","Implement", "Cast", "Throw", "Annotation")
+
+    # Get all the parameter values for a specific cell
+    for (j in 1: length(parameters)) {
+      current_param <- depends_table[i][[parameters[[j]]]]
+      if (current_param > 0){
+        values[[parameters[[j]]]] <- current_param
+      }
+    }
+
+    return(list(src=src_index, dest=dest_index, values=data.frame(values)))
+  }
+
+  # Get the cells
+  cells <- lapply(cells_indices, getCell)
+  cells_df <- data.table::data.table(jsonlite::fromJSON(jsonlite::toJSON(cells, auto_unbox = TRUE)))
+
+  # Sort by the src_index and dest_index if is_sorted is true
+  if (is_sorted==TRUE){
+    data.table::setorder(cells_df, cols = "src", "dest")
+  }
+
+  sdsm_json <- list(schemaVersion="1.0", name=paste0(project_name, "-sdsm"), variables=variables, cells=cells_df)
+
+  json_df <- jsonlite::fromJSON(jsonlite::toJSON(sdsm_json, auto_unbox = TRUE))
+
+  # Unbox each of the cell values (should be values: object, not values: [object])
+  json_df$cells$values <- lapply(json_df$cells$values, jsonlite::unbox)
+
+  # Save the json to a file
+  jsonlite::write_json(json_df, paste0("../rawdata/dv8/", project_name, ".json"), auto_unbox=TRUE)
+}
+
+#' Convert gitlog table to a hdsm.json
+#'
+#' Converts a gitlog table into an *-hdsm.json binary.
+#'
+#' @param gitlog_table parsed gitlog table
+#' @param project_name project name for the hdsm.json
+#' @param is_sorted whether the json is sorted by src and dest
+#' @export
+#' @family dv8
+gitlog_to_hdsmj <- function(gitlog_table, project_name, is_sorted=FALSE){
+  # Call preliminary functions to get graph and cochange for the files
+  gitlog_graph <- transform_gitlog_to_bipartite_network(gitlog_table, mode ="commit-file")
+  cochange_table <- bipartite_graph_projection(gitlog_graph, mode = FALSE, is_intermediate_projection = FALSE)
+
+  # List of two tables. 2nd table had the filenames and cochange
+  cochange_table <- cochange_table[[2]]
+
+  # Get the files in the from (src) and to (dest) columns, concatenate them into one list, sort alphabetically
+  # Getting files in the "from" column
+  gitlog_files_src <- unique(cochange_table[["from"]])
+  # Getting files in the "to" column
+  gitlog_files_dest <- unique(cochange_table[["to"]])
+  # Combining src and dest columns
+  gitlog_files <- c(gitlog_files_src, gitlog_files_dest)
+
+  # Taking the unique combined columns
+  variables <- unique(gitlog_files)
```

Comment:
@leilani-reich This, paired with your assumption of line 172, 

```r
 # List of two tables. 2nd table had the filenames and cochange
  cochange_table <- cochange_table[[2]]
```

is where the second part of the "variables are missing" is introduced in #184. More specifically, variables should be pulled from `cochange_table[[1]]`. I'd also recommend you rename these variables here for clarity:

cochange_table[[1]] should be referred to as `nodes_table`, where as `cochange_table[[2]]` should be referred to as `edgelist_table`.

Here's what `nodes_table` looks like, after the fix of part 1 in #184 (not committed yet at the time of this comment):

```
                        name  type   color
 1:              .gititgnore FALSE #f4dbb5
 2:                README.md FALSE #f4dbb5
 3:  QtNotepad/QtNotepad.pro FALSE #f4dbb5
 4:       QtNotepad/main.cpp FALSE #f4dbb5
 5: QtNotepad/mainwindow.cpp FALSE #f4dbb5
 6:   QtNotepad/mainwindow.h FALSE #f4dbb5
 7:  QtNotepad/mainwindow.ui FALSE #f4dbb5
 8:  QtNotepad/resources.qrc FALSE #f4dbb5
 9:               .gitignore FALSE #f4dbb5
10:                  LICENSE FALSE #f4dbb5
11:               LICENSE.md FALSE #f4dbb5
```

As you can see, the all the files are listed here. 

Let me clarify what structure you are looking at for "cochange_table": When a graph (the circles with connections of your Social Smell cheatsheet) is represented in memory, you have a few options. One of them is representing them as a pair of tables: Nodes, which defines all information a node will have (e.g. the color of a node, it's size, etc), 

and an edgelist table. This edgelist table defines only the connections, which happens in **pairs**. In this code, you assumed that the edgelist contains all the nodes, but that assumption is not always true. 

Indeed, in this project, LICENSE.md is committed only once (during a file rename), and more importantly, it is committed without any other file. This leads to LICENSE.md having no connections to any other file in the entire of this git log, and thus will not appear on the edgelist table. **But it will appear on the node table**. 

This format of node and edgelist tables is also what DV8 is using: Note in the file you sent me (in this comment https://github.com/sailuh/kaiaulu/issues/184#issue-1662363636), LICENSE.md appears under "variables" (DV8's node list table), but not in the "cells" (DV8's edgelist table).